File: server/simple-ocr.ts
import { spawn } from 'child_process';
import path from 'path';
import fs from 'fs';
import { promisify } from 'util';

const writeFile = promisify(fs.writeFile);
const unlink = promisify(fs.unlink);

interface SimpleOCRResult {
  success: boolean;
  raw_text?: string;
  text_blocks?: Array<{
    text: string;
    confidence: number;
    x: number;
    y: number;
    width: number;
    height: number;
  }>;
  total_blocks?: number;
  error?: string;
  method: string;
}

/**
 * Simple Pure OCR - Just reads text from image without complex interpretation
 * Uses temporary file to avoid command line argument length limits
 */
export async function analyzeImageWithSimpleOCR(imageBase64: string): Promise<SimpleOCRResult> {
  let tempImagePath: string | null = null;
  
  try {
    console.log('Starting simple pure OCR processing...');
    
    // Create temporary image file
    const imageBuffer = Buffer.from(imageBase64, 'base64');
    tempImagePath = path.join(process.cwd(), 'temp', `ocr_image_${Date.now()}_${Math.random().toString(36).substr(2, 9)}.png`);
    
    // Ensure temp directory exists
    const tempDir = path.dirname(tempImagePath);
    if (!fs.existsSync(tempDir)) {
      fs.mkdirSync(tempDir, { recursive: true });
    }
    
    // Write image to temporary file
    await writeFile(tempImagePath, imageBuffer);
    console.log('Image saved to temporary file:', tempImagePath);
    
    // Run Python OCR script
    const pythonPath = 'python3';
    const scriptPath = path.join(process.cwd(), 'server', 'simple_ocr.py');
    
    console.log('Running Python OCR script:', scriptPath);
    
    return new Promise((resolve, reject) => {
      const pythonProcess = spawn(pythonPath, [scriptPath], {
        stdio: ['pipe', 'pipe', 'pipe']
      });
      
      let stdout = '';
      let stderr = '';
      
      pythonProcess.stdout.on('data', (data) => {
        stdout += data.toString();
      });
      
      pythonProcess.stderr.on('data', (data) => {
        stderr += data.toString();
      });
      
      pythonProcess.on('close', async (code) => {
        // Clean up temporary file
        if (tempImagePath && fs.existsSync(tempImagePath)) {
          try {
            await unlink(tempImagePath);
            console.log('Temporary file cleaned up');
          } catch (cleanupError) {
            console.warn('Failed to clean up temporary file:', cleanupError);
          }
        }
        
        if (code !== 0) {
          console.error('Simple OCR process failed with code:', code);
          console.error('Simple OCR stderr:', stderr);
          reject(new Error(`Simple OCR failed with code ${code}: ${stderr}`));
          return;
        }
        
        try {
          const result = JSON.parse(stdout);
          
          if (!result.success) {
            console.error('Simple OCR extraction failed:', result.error);
            reject(new Error(result.error));
            return;
          }
          
          console.log(`Simple OCR extraction completed successfully`);
          console.log(`Extracted ${result.total_blocks || 0} text blocks`);
          
          resolve(result);
          
        } catch (parseError) {
          console.error('Failed to parse Simple OCR result:', parseError);
          console.error('Raw stdout:', stdout);
          reject(parseError);
        }
      });
      
      pythonProcess.on('error', async (error) => {
        // Clean up temporary file on error
        if (tempImagePath && fs.existsSync(tempImagePath)) {
          try {
            await unlink(tempImagePath);
          } catch (cleanupError) {
            console.warn('Failed to clean up temporary file on error:', cleanupError);
          }
        }
        console.error('Simple OCR process error:', error);
        reject(error);
      });
      
      // Send image path to Python script via stdin
      pythonProcess.stdin.write(tempImagePath + '\n');
      pythonProcess.stdin.end();
    });
    
  } catch (error) {
    // Clean up temporary file on error
    if (tempImagePath && fs.existsSync(tempImagePath)) {
      try {
        await unlink(tempImagePath);
      } catch (cleanupError) {
        console.warn('Failed to clean up temporary file on error:', cleanupError);
      }
    }
    throw error;
  }
}

File: server/simple_ocr.py
#!/usr/bin/env python3
"""
Simple Pure OCR - Just reads text from image without complex parsing
Uses file input instead of command line arguments to avoid size limits
"""

import sys
import json
import base64
import tempfile
import os
from PIL import Image
import pytesseract
from typing import Dict, Any, List

def extract_pure_text_from_file(image_path: str) -> Dict[str, Any]:
    """
    Pure OCR extraction from image file - just reads all text
    Returns raw text without interpretation
    """
    try:
        # Open and process image
        image = Image.open(image_path)
        
        # Convert to RGB if necessary
        if image.mode != 'RGB':
            image = image.convert('RGB')
        
        # Simple OCR with Portuguese support
        custom_config = r'--oem 3 --psm 6 -l por+eng'
        
        # Extract all text
        raw_text = pytesseract.image_to_string(image, config=custom_config)
        
        # Extract text with bounding boxes for verification
        data = pytesseract.image_to_data(image, config=custom_config, output_type=pytesseract.Output.DICT)
        
        # Build text blocks with positions for user verification
        text_blocks = []
        for i in range(len(data['text'])):
            text = data['text'][i].strip()
            if text and int(data['conf'][i]) > 30:  # Only confident text
                text_blocks.append({
                    'text': text,
                    'confidence': int(data['conf'][i]),
                    'x': data['left'][i],
                    'y': data['top'][i],
                    'width': data['width'][i],
                    'height': data['height'][i]
                })
        
        # Sort blocks by position (top to bottom, left to right)
        text_blocks.sort(key=lambda block: (block['y'], block['x']))
        
        return {
            'success': True,
            'raw_text': raw_text.strip(),
            'text_blocks': text_blocks,
            'total_blocks': len(text_blocks),
            'method': 'simple_pure_ocr'
        }
        
    except Exception as e:
        return {
            'success': False,
            'error': str(e),
            'method': 'simple_pure_ocr'
        }

def main():
    """Command line interface - reads image path from stdin"""
    try:
        # Read image path from stdin
        image_path = input().strip()
        
        if not os.path.exists(image_path):
            print(json.dumps({'success': False, 'error': f'Image file not found: {image_path}'}))
            sys.exit(1)
        
        result = extract_pure_text_from_file(image_path)
        print(json.dumps(result, ensure_ascii=False, indent=2))
        
    except Exception as e:
        print(json.dumps({'success': False, 'error': str(e)}))
        sys.exit(1)

if __name__ == '__main__':
    main()